massa-execution-worker/src/speculative_async_pool.rs

```diff
@@ -154,20 +154,24 @@ impl SpeculativeAsyncPool {
         // Filter out all messages for which the validity end is expired.
         // Note that the validity_end bound is NOT included in the validity interval of the message.
 
-        let mut eliminated_infos: Vec<_> = self
+        let (mut eliminated_infos, kept_infos) = self
             .message_infos
-            .extract_if(|_k, v| *slot >= v.validity_end)
-            .collect();
-
-        let eliminated_new_messages: Vec<_> = self
-            .pool_changes
-            .0
-            .extract_if(|_k, v| match v {
-                SetUpdateOrDelete::Set(v) => *slot >= v.validity_end,
-                SetUpdateOrDelete::Update(_v) => false,
-                SetUpdateOrDelete::Delete => false,
-            })
-            .collect();
+            .clone()
+            .into_iter()
+            .partition::<Vec<_>, _>(|(_id, info)| *slot >= info.validity_end);
+        self.message_infos = kept_infos.into_iter().collect();
```

Comment:
this collect is heavy (eg. it does deduplication). Better not change the type above, use the mem::take, and just reaffact: self.message_infos = kept_infos.